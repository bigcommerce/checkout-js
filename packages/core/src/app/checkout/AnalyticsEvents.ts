let boltTracker: Window["BoltTrack"] = {
  recordEvent: (event: string) => console.log("--bolt bigc--: NOOP ", event),
};

const eventLog: string[] = [];

export const AnalyticsEvents = {
  init: () => {
    try {
      console.log("--bolt bigc--: intializing Bolt-BigC analytics events");
      if (window && window.BoltTrack) {
        boltTracker = window.BoltTrack;
        console.log("--bolt bigc--: successfully assigned window BoltTrack");
        // immediately send checkout load success event
        AnalyticsEvents.emitEvent("Checkout load success");
      }
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  emitEvent: (eventName: string) => {
    try {
      console.log("--bolt bigc--: emitting analytics event ", eventName);
      const props: any = {
        nextState: eventName,
        prevState: "",
      };
      if (eventLog.length > 0) {
        props.prevState = eventLog[0];
      }
      boltTracker.recordEvent("CheckoutFunnelTransition", props);
      // add latest event to beginning of log array
      eventLog.unshift(eventName);
    } catch (e) {
      console.log("--bolt bigc--: error during emitting event ", e);
    }
  },
  onBeforeUnload: () => {
    AnalyticsEvents.emitEvent("Exit");
  },
};
