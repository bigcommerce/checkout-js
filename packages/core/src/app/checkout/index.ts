export { RenderCheckout, RenderCheckoutOptions } from './renderCheckout';
export { default as CheckoutContext, CheckoutContextProps } from './CheckoutContext';
export {
    default as CheckoutProvider,
    CheckoutProviderProps,
    CheckoutProviderState,
} from './CheckoutProvider';
export { default as withCheckout, WithCheckoutProps } from './withCheckout';
export { default as CheckoutSupport } from './CheckoutSupport';
export { default as NoopCheckoutSupport } from './NoopCheckoutSupport';
export { default as navigateToOrderConfirmation } from './navigateToOrderConfirmation';
