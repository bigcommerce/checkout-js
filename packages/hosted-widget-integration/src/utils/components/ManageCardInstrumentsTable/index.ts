export { default as ManageCardInstrumentsTable } from './ManageCardInstrumentsTable';
