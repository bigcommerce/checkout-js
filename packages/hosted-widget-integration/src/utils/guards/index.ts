export { default as isAccountInstrument } from './is-account-instrument';
export { default as assertIsCardInstrument, isCardInstrument } from './is-card-instrument';
export { default as assertIsBankInstrument, isBankAccountInstrument } from './is-bank-instrument';
