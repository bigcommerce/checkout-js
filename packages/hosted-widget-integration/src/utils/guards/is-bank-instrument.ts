import { BankInstrument, PaymentInstrument } from '@bigcommerce/checkout-sdk';

export function isBankAccountInstrument(
    instrument?: PaymentInstrument,
): instrument is BankInstrument {
    return instrument !== undefined && instrument.type === 'bank';
}

export default function assertIsBankInstrument(
    instrument?: PaymentInstrument,
): asserts instrument is BankInstrument {
    if (!isBankAccountInstrument(instrument)) {
        throw new Error('Is not a bank account instrument.');
    }
}
