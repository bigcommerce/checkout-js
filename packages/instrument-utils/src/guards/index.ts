/* istanbul ignore file */
export { isAccountInstrument } from './isAccountInstrument';
export { isAchInstrument } from './isAchInstrument';
export { assertIsBankInstrument, isBankAccountInstrument } from './isBankInstrument';
export { assertIsCardInstrument, isCardInstrument } from './isCardInstrument';
export { isInstrumentCardCodeRequired } from './isInstrumentCardCodeRequired';
export { isInstrumentFeatureAvailable } from './isInstrumentFeatureAvailable';
export {
    isInstrumentCardNumberRequired,
    IsInstrumentCardNumberRequiredState,
} from './isInstrumentCardNumberRequired';
export { default as isInstrumentCardCodeRequiredSelector } from './isInstrumentCardCodeRequiredSelector';
export { default as isInstrumentCardNumberRequiredSelector } from './isInstrumentCardNumberRequiredSelector';
