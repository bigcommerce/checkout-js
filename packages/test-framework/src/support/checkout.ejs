<!DOCTYPE html>
<html>
    <head>
        <title>Local Testing Environment</title>
        <base href="/" target="_top"/>
        <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />
        <link href="https://fonts.googleapis.com/css?family=Karla:400|Montserrat:400,500,700&display=swap" rel="stylesheet" />
    </head>
    <body>
    <header class="checkoutHeader optimizedCheckout-header">
        <div class="checkoutHeader-content">
            <h1 class="is-srOnly">Checkout</h1>
            <h2 class="checkoutHeader-heading">
                <a class="checkoutHeader-link" href="<%= storeUrl %>/">
                    <span class="header-logo-text">Local Testing Environment</span>
                </a>
            </h2>
        </div>
    </header>
    <div class="ng-checkout-container" id="micro-app-ng-checkout">
        <div class="ng-checkout-view" data-ui-view></div>
    </div>
    <script type="text/javascript">
        window.checkoutConfig = {
            checkoutId: '<%= checkoutId %>',
            containerId: 'micro-app-ng-checkout',
        };
    </script>
    <script src="<%= localhostUrl %>/auto-loader.js" type="text/javascript"></script>
    <script type="text/javascript">
        (function() {
            window.checkout.renderCheckout({
                checkoutId: '<%= checkoutId %>',
                containerId: 'micro-app-ng-checkout',
                sentryConfig: null,
            })
        })();
    </script>
    </body>
</html>
