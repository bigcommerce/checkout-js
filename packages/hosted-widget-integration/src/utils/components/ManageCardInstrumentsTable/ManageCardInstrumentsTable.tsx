import { CardInstrument } from '@bigcommerce/checkout-sdk';
import { expirationDate } from 'card-validator';
import classNames from 'classnames';
import creditCardType from 'credit-card-type';
import React, { FunctionComponent, memo, useCallback } from 'react';

import { TranslatedString } from '@bigcommerce/checkout/locale';
import { CreditCardIcon, LoadingOverlay } from '@bigcommerce/checkout/ui';

function mapFromInstrumentCardType(type: string): string {
    switch (type) {
        case 'amex':
        case 'american_express':
            return 'american-express';

        case 'diners':
            return 'diners-club';

        default:
            return type;
    }
}

interface ManageInstrumentsRowProps {
    instrument: CardInstrument;
    onDeleteInstrument(id: string): void;
}

const ManageInstrumentsRow: FunctionComponent<ManageInstrumentsRowProps> = ({
    instrument,
    onDeleteInstrument,
}) => {
    const cardType = mapFromInstrumentCardType(instrument.brand);
    const cardInfo = creditCardType.getTypeInfo(cardType);
    const isExpired =
        expirationDate({
            month: instrument.expiryMonth,
            year: instrument.expiryYear,
        }).isValid === false;

    const handleDelete = useCallback(() => {
        onDeleteInstrument(instrument.bigpayToken);
    }, [instrument, onDeleteInstrument]);

    return (
        <tr>
            <td data-test="manage-instrument-cardType">
                <CreditCardIcon cardType={cardType} />

                {
                    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                    cardInfo && (
                        <span className="instrumentModal-instrumentCardType">
                            {cardInfo.niceType}
                        </span>
                    )
                }
            </td>
            <td data-test="manage-instrument-last4">{instrument.last4}</td>
            <td
                className={classNames({ 'instrumentModal-instrumentExpiry--expired': isExpired })}
                data-test="manage-instrument-expiry"
            >
                {`${instrument.expiryMonth}/${instrument.expiryYear}`}
            </td>
            <td>
                <button
                    className="button button--tiny table-actionButton optimizedCheckout-buttonSecondary"
                    data-test="manage-instrument-delete-button"
                    onClick={handleDelete}
                    type="button"
                >
                    <TranslatedString id="common.delete_action" />
                </button>
            </td>
        </tr>
    );
};

export interface ManageCardInstrumentsTableProps {
    instruments: CardInstrument[];
    isDeletingInstrument: boolean;
    onDeleteInstrument(id: string): void;
}

const ManageCardInstrumentsTable: FunctionComponent<ManageCardInstrumentsTableProps> = ({
    instruments,
    isDeletingInstrument,
    onDeleteInstrument,
}) => {
    if (instruments.length === 0) {
        return (
            <p>
                <TranslatedString id="payment.instrument_manage_modal_empty_text" />
            </p>
        );
    }

    return (
        <LoadingOverlay isLoading={isDeletingInstrument}>
            <table className="table">
                <thead className="table-thead">
                    <tr>
                        <th>
                            <TranslatedString id="payment.instrument_manage_table_header_payment_method_text" />
                        </th>
                        <th>
                            <TranslatedString id="payment.instrument_manage_table_header_ending_in_text" />
                        </th>
                        <th>
                            <TranslatedString id="payment.instrument_manage_table_header_expiry_date_text" />
                        </th>
                        <th />
                    </tr>
                </thead>

                <tbody className="table-tbody">
                    {instruments.map((instrument) => (
                        <ManageInstrumentsRow
                            instrument={instrument}
                            key={instrument.bigpayToken}
                            onDeleteInstrument={onDeleteInstrument}
                        />
                    ))}
                </tbody>
            </table>
        </LoadingOverlay>
    );
};

export default memo(ManageCardInstrumentsTable);
