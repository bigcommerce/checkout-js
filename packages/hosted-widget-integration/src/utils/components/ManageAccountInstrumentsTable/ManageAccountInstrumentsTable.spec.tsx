import { mount, render } from 'enzyme';
import React from 'react';

import {
    createLocaleContext,
    LocaleContext,
    LocaleContextType,
} from '@bigcommerce/checkout/locale';
import { getInstruments, getStoreConfig } from '@bigcommerce/checkout/test-utils';
import { LoadingOverlay } from '@bigcommerce/checkout/ui';

import { isAccountInstrument } from '../../guards';

import ManageCardInstrumentsTable, {
    ManageAccountInstrumentsTableProps,
} from './ManageAccountInstrumentsTable';

describe('ManageAccountInstrumentsTable', () => {
    let defaultProps: ManageAccountInstrumentsTableProps;
    let localeContext: LocaleContextType;

    beforeEach(() => {
        defaultProps = {
            instruments: getInstruments().filter(isAccountInstrument),
            isDeletingInstrument: false,
            onDeleteInstrument: jest.fn(),
        };

        localeContext = createLocaleContext(getStoreConfig());
    });

    it('matches snapshot with rendered output', () => {
        const component = render(
            <LocaleContext.Provider value={localeContext}>
                <ManageCardInstrumentsTable {...defaultProps} />
            </LocaleContext.Provider>,
        );

        expect(component).toMatchSnapshot();
    });

    it('renders instrument as row in table', () => {
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <ManageCardInstrumentsTable {...defaultProps} />
            </LocaleContext.Provider>,
        );

        expect(
            component.find('[data-test="manage-instrument-accountExternalId"]').at(0).text(),
        ).toBe('test@external-id.com');
    });

    it('triggers callback when delete button is clicked', () => {
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <ManageCardInstrumentsTable {...defaultProps} />
            </LocaleContext.Provider>,
        );

        component.find('[data-test="manage-instrument-delete-button"]').at(1).simulate('click');

        expect(defaultProps.onDeleteInstrument).toHaveBeenCalledWith(
            defaultProps.instruments[1].bigpayToken,
        );
    });

    it('renders message if there are no available instruments', () => {
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <ManageCardInstrumentsTable {...defaultProps} instruments={[]} />
            </LocaleContext.Provider>,
        );

        expect(component.text()).toEqual(
            localeContext.language.translate('payment.instrument_manage_modal_empty_text'),
        );
    });

    it('shows loading overlay when deleting', () => {
        const component = mount(
            <LocaleContext.Provider value={localeContext}>
                <ManageCardInstrumentsTable {...defaultProps} isDeletingInstrument={true} />
            </LocaleContext.Provider>,
        );

        expect(component.find(LoadingOverlay).prop('isLoading')).toBe(true);
    });
});
