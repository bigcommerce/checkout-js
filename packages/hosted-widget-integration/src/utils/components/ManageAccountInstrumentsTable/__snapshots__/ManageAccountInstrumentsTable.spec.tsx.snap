// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageAccountInstrumentsTable matches snapshot with rendered output 1`] = `
<div
  class="loadingOverlay-container"
>
  <table
    class="table"
  >
    <thead
      class="table-thead"
    >
      <tr>
        <th>
          Payment method
        </th>
        <th />
      </tr>
    </thead>
    <tbody
      class="table-tbody"
    >
      <tr>
        <td
          data-test="manage-instrument-accountExternalId"
        >
          <div
            class="icon accountIcon-icon icon--medium"
          >
            <svg
              height="300"
              viewBox="0 0 255 300"
              width="255"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M228.03 77.2c3.6-23.02-.03-38.7-12.48-52.89C201.83 8.7 177.07 2 145.4 2H53.4c-6.47 0-11.99 4.7-12.99 11.11L2.1 256.03a7.9 7.9 0 0 0 7.8 9.14h56.78c79.33-53.38 148.45-81.4 161.35-187.97z"
                fill="#253B80"
              />
              <path
                d="M228.02 77.2C124 73 95 98 80.5 177.52l-13.82 87.65-3.91 24.85a6.9 6.9 0 0 0 6.82 7.98h47.86c5.67 0 10.48-4.12 11.37-9.71l.48-2.43 9-57.2.59-3.15c.88-5.6 5.7-9.73 11.37-9.73h7.16c46.38 0 82.68-18.83 93.29-73.33 4.43-22.76 2.14-41.77-9.6-55.14a45.77 45.77 0 0 0-13.1-10.1z"
                fill="#179BD7"
              />
              <path
                d="M215.33 72.14c-3.88-1.12-7.82-2-11.8-2.62a149.83 149.83 0 0 0-23.79-1.73h-72.1a11.49 11.49 0 0 0-11.36 9.73l-15.34 97.16-.44 2.84a13.1 13.1 0 0 1 12.95-11.1h27c53 0 94.5-21.53 106.62-83.82.37-1.85.67-3.64.95-5.4a64.66 64.66 0 0 0-9.98-4.2c-.9-.3-1.8-.59-2.71-.86z"
                fill="#222D65"
              />
            </svg>
          </div>
          <span
            class="instrumentModal-instrumentAccountExternalId"
          >
            test@external-id.com
          </span>
        </td>
        <td>
          <button
            class="button button--tiny table-actionButton optimizedCheckout-buttonSecondary"
            data-test="manage-instrument-delete-button"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td
          data-test="manage-instrument-accountExternalId"
        >
          <div
            class="icon accountIcon-icon icon--medium"
          >
            <svg
              height="300"
              viewBox="0 0 255 300"
              width="255"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M228.03 77.2c3.6-23.02-.03-38.7-12.48-52.89C201.83 8.7 177.07 2 145.4 2H53.4c-6.47 0-11.99 4.7-12.99 11.11L2.1 256.03a7.9 7.9 0 0 0 7.8 9.14h56.78c79.33-53.38 148.45-81.4 161.35-187.97z"
                fill="#253B80"
              />
              <path
                d="M228.02 77.2C124 73 95 98 80.5 177.52l-13.82 87.65-3.91 24.85a6.9 6.9 0 0 0 6.82 7.98h47.86c5.67 0 10.48-4.12 11.37-9.71l.48-2.43 9-57.2.59-3.15c.88-5.6 5.7-9.73 11.37-9.73h7.16c46.38 0 82.68-18.83 93.29-73.33 4.43-22.76 2.14-41.77-9.6-55.14a45.77 45.77 0 0 0-13.1-10.1z"
                fill="#179BD7"
              />
              <path
                d="M215.33 72.14c-3.88-1.12-7.82-2-11.8-2.62a149.83 149.83 0 0 0-23.79-1.73h-72.1a11.49 11.49 0 0 0-11.36 9.73l-15.34 97.16-.44 2.84a13.1 13.1 0 0 1 12.95-11.1h27c53 0 94.5-21.53 106.62-83.82.37-1.85.67-3.64.95-5.4a64.66 64.66 0 0 0-9.98-4.2c-.9-.3-1.8-.59-2.71-.86z"
                fill="#222D65"
              />
            </svg>
          </div>
          <span
            class="instrumentModal-instrumentAccountExternalId"
          >
            test@external-id-2.com
          </span>
        </td>
        <td>
          <button
            class="button button--tiny table-actionButton optimizedCheckout-buttonSecondary"
            data-test="manage-instrument-delete-button"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
